protocols/tls13/src/tls13handshake.rs

```diff
@@ -0,0 +1,491 @@
+    let sender_write_key_iv = derive_aead_key_iv(ha, ae, &client_early_traffic_secret)?;
+    Ok((sender_write_key_iv, early_exporter_master_secret))
+}
+
+pub fn derive_finished_key(ha: &HashAlgorithm, k: &KEY) -> Res<MACK> {
+    Ok(hkdf_expand_label(ha,k,&bytes(&label_finished),&empty(),hmac_key_len(ha))?)
+}
+
+pub fn derive_hk_ms(
+    ha: &HashAlgorithm,
+    ae: &AEADAlgorithm,
+    gxy: &KEY,
+    psko: &Option<PSK>,
+    tx: &HASH,
+) -> Res<(AEKIV, AEKIV, MACK, MACK, KEY)> {
+    let psk = if let Some(k) = psko {KEY::from_seq(k)} else {zero_key(ha)};
+    let early_secret = hkdf_extract(ha, &psk, &zero_key(ha))?;
+    let handshake_secret = hkdf_extract(ha, gxy, &early_secret)?;
+    let client_handshake_traffic_secret =
+        derive_secret(ha, &handshake_secret, &bytes(&label_c_hs_traffic), tx)?;
+    let server_handshake_traffic_secret =
+        derive_secret(ha, &handshake_secret, &bytes(&label_s_hs_traffic), tx)?;
+    let client_finished_key = derive_finished_key(ha, &client_handshake_traffic_secret)?;
+    let server_finished_key = derive_finished_key(ha, &server_handshake_traffic_secret)?;
+    let client_write_key_iv = derive_aead_key_iv(ha, ae, &client_handshake_traffic_secret)?;
+    let server_write_key_iv = derive_aead_key_iv(ha, ae, &server_handshake_traffic_secret)?;
+    let master_secret_ =
+        derive_secret(ha, &handshake_secret, &bytes(&label_derived), &hash_empty(ha)?)?;
+    let master_secret = hkdf_extract(ha, &zero_key(ha), &master_secret_)?;
+    Ok((
+        client_write_key_iv,
+        server_write_key_iv,
+        client_finished_key,
+        server_finished_key,
+        master_secret,
+    ))
+}
+
+pub fn derive_app_keys(
+    ha: &HashAlgorithm,
+    ae: &AEADAlgorithm,
+    master_secret: &KEY,
+    tx: &HASH,
+) -> Res<(AEKIV, AEKIV, KEY)> {
+    let client_application_traffic_secret_0 =
+        derive_secret(ha, &master_secret, &bytes(&label_c_ap_traffic), tx)?;
+    let server_application_traffic_secret_0 =
+        derive_secret(ha, &master_secret, &bytes(&label_s_ap_traffic), tx)?;
+    let client_write_key_iv = derive_aead_key_iv(ha, ae, &client_application_traffic_secret_0)?;
+    let server_write_key_iv = derive_aead_key_iv(ha, ae, &server_application_traffic_secret_0)?;
+    let exporter_master_secret = derive_secret(ha, master_secret, &bytes(&label_exp_master), tx)?;
+    Ok((
+        client_write_key_iv,
+        server_write_key_iv,
+        exporter_master_secret,
+    ))
+}
+
+pub fn derive_rms(ha: &HashAlgorithm, master_secret: &KEY, tx: &HASH) -> Res<KEY> {
+    let resumption_master_secret = derive_secret(ha, master_secret, &bytes(&label_res_master), tx)?;
+    Ok(resumption_master_secret)
+}
+
+
+/* Incremental Transcript Construction 
+   For simplicity, we store the full transcript, but an internal hash state would suffice. */
+
+pub struct TranscriptTruncatedClientHello(pub HashAlgorithm, pub HASH);
```

Comment:
Can be viewed as a one-constructor sum type, so I guess that's OK.

Reply:
Yep, these are quite useful. We can also erase them to (unnamed) tuples, but we get better errors with this abstraction.